packages/neon/neon_talk/lib/src/widgets/message_input.dart

```diff
@@ -0,0 +1,197 @@
+import 'package:neon_talk/src/blocs/room.dart';
+import 'package:nextcloud/spreed.dart' as spreed;
+
+/// Widget for displaying the emoji button, text input and send button.
+class TalkMessageInput extends StatefulWidget {
+  /// Creates a new Talk message input.
+  const TalkMessageInput({
+    super.key,
+  });
+
+  @override
+  State<TalkMessageInput> createState() => _TalkMessageInputState();
+}
+
+class _TalkMessageInputState extends State<TalkMessageInput> {
+  final controller = TextEditingController();
+  final focusNode = FocusNode();
+  late TalkRoomBloc bloc;
+
+  @override
+  void initState() {
+    super.initState();
+
+    bloc = NeonProvider.of<TalkRoomBloc>(context);
+  }
+
+  @override
+  void dispose() {
+    controller.dispose();
+    focusNode.dispose();
+    super.dispose();
+  }
+
+  void sendMessage() {
+    final message = controller.text;
+    if (message.isNotEmpty) {
+      bloc.sendMessage(message);
+      controller.clear();
+    }
+    focusNode.requestFocus();
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    Widget? emojiButton;
+    // On cupertino the keyboard always has an option to insert emojis, so we don't need to add it
+    if (!isCupertino(context)) {
+      emojiButton = IconButton(
+        tooltip: TalkLocalizations.of(context).roomMessageAddEmoji,
+        onPressed: () async {
+          final emoji = await showDialog<String>(
+            context: context,
+            builder: (context) => const NeonEmojiPickerDialog(),
+          );
+          if (emoji != null) {
+            final text = controller.text;
+            final textSelection = controller.selection;
+
+            controller
+              ..text = text.replaceRange(textSelection.start, textSelection.end, emoji)
+              ..selection = textSelection.copyWith(
+                baseOffset: textSelection.start + emoji.length,
+                extentOffset: textSelection.start + emoji.length,
+              );
+          }
+        },
+        icon: const Icon(Icons.emoji_emotions_outlined),
+      );
+    }
+
+    return TypeAheadField<_Suggestion>(
+      direction: VerticalDirection.up,
+      hideOnEmpty: true,
+      debounceDuration: const Duration(milliseconds: 50),
+      controller: controller,
+      focusNode: focusNode,
+      suggestionsCallback: (_) async {
+        final cursor = controller.selection.start;
+        if (controller.text.isEmpty || cursor != controller.selection.end) {
+          return [];
+        }
+
+        String? matchingPart;
+        final parts = controller.text.split(' ');
+        var skippedChars = 0;
+        for (final part in parts) {
+          if (skippedChars <= cursor && (skippedChars + part.length + 1) >= cursor) {
+            matchingPart = part;
+            break;
+          }
+          skippedChars += part.length + 1;
+        }
+
+        if (matchingPart == null || !matchingPart.startsWith('@') || matchingPart.length <= 1) {
+          return [];
+        }
```

Comment:
Quite a few iterations and plits
```suggestion
        String? matchingPart = controller.text.substring(0, cursor);
        final index = matchingPart.lastIndexOf(' ') + 1;
        matchingPart = matchingPart.substring(index);
        if (!matchingPart.startsWith('@') || matchingPart.length <= 1) {
          return [];
        }
```

In theory we can also do the condition without spliting the string again (character at index (if existing) is @ and the difference between the cursor and index is greater than 2 but that's hard to understand later)

Reply:
Thanks, this makes perfect sense. I always forget that there are so many useful methods in Dart :sweat: